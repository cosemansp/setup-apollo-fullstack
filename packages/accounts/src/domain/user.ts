export interface User {
  id: number;
  firstName: string;
  lastName: string;
  age: number;
  email: string;
  image: string;
  phone: string;
  company: string;
  address: {
    street: string;
    city: string;
    zip: string;
  };
}
