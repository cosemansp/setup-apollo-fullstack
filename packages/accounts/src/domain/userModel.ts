import { Document, Schema, Model, model } from 'mongoose';

export interface Address {
  street: string;
  city: string;
  zip: string;
}

export interface User {
  name?: string;
  age?: number;
  email?: string;
  image?: string;
  phone?: string;
  company?: string;
  //  address: Address;
}

export interface UserDoc extends User, Document {}

export var UserSchema: Schema = new Schema({
  name: String,
  email: String,
  image: String,
  phone: String,
  company: String,
});

export type UserModelType = Model<UserDoc>;
export const UserModel: UserModelType = model<UserDoc>('User', UserSchema);
