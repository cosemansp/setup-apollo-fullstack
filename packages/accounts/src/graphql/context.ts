import { Request, Response } from 'express';
import { UserDataSource } from '../dataSources/userDataSource';

export interface JWTPayload {
  sub: string;
  email?: string;
  iat?: number;
  exp?: number;
}

export interface User extends JWTPayload {}

export interface Context {
  req: Request;
  res: Response;
  user?: User;
  dataSources: {
    user: UserDataSource;
  };
}
