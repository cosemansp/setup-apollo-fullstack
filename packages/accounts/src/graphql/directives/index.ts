export * from './hasRoleDirective';
export * from './hasScopeDirective';
export * from './isAuthDirective';
