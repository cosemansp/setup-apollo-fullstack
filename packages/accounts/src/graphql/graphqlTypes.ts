import { MutationResolvers } from './types';

export * from './types';

export type RegisterMutation = MutationResolvers['register'];
export type LoginMutation = MutationResolvers['login'];
export type UpdateUserMutation = MutationResolvers['updateUser'];
export type RemoveUserMutation = MutationResolvers['removeUser'];
