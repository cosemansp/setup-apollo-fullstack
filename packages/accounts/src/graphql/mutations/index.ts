export * from './login';
export * from './register';
export * from './updateUser';
export * from './removeUser';
