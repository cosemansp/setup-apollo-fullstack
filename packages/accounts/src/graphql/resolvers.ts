import { Resolvers } from './graphqlTypes';
import { register, login, updateUser, removeUser } from './mutations';

export const resolvers: Resolvers = {
  Query: {
    //
    // query me
    //
    me(root, args, { dataSources, user }) {
      // check if user is available (valid token)
      if (!user) {
        return null;
      }

      // retrieve user
      return dataSources.user.loadOne(user && user.sub);
    },

    //
    // query user
    //
    user(root, args, { dataSources }) {
      return dataSources.user.loadOne(args.id);
    },

    //
    // query users
    //
    users(root, args, { dataSources }) {
      return dataSources.user.loadManyByQuery({});
    },
  },
  LoginPayload: {
    __resolveType(obj) {
      return obj.__typename;
    },
  },
  RegisterPayload: {
    __resolveType(obj) {
      return obj.__typename;
    },
  },
  Mutation: {
    register,
    login,
    updateUser,
    removeUser,
  },
};
