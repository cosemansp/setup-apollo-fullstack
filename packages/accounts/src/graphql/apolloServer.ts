import util from 'util';
import { ApolloServer } from 'apollo-server-express';
import { buildFederatedSchema } from '@apollo/federation';
import { GraphQLSchema } from 'graphql';
import { verify, decode } from 'jsonwebtoken';
import { makeExecutableSchema } from 'apollo-server';
import { resolvers } from './resolvers';
import { typeDefs } from './schema';
import { Context, JWTPayload } from './context';
import { logManager } from '../logManager';
import { UserDataSource } from '@/dataSources/userDataSource';
import config from '@/config';

const log = logManager.getLogger('root.graphql');

// create GraphqlSchema
let graphqlSchema: GraphQLSchema;
if (config.FEDERATED) {
  graphqlSchema = buildFederatedSchema([{ typeDefs, resolvers }]);
} else {
  graphqlSchema = makeExecutableSchema({
    typeDefs,
    resolvers,
    allowUndefinedInResolve: true,
  });
}

// we need to omit the dataSources because they are added later by the graphqlServer
type BaseContext = Omit<Context, 'dataSources'>;

export const server = new ApolloServer({
  schema: graphqlSchema,
  context: ({ res, req }): BaseContext => {
    // Base context
    const context: BaseContext = {
      req,
      res,
    };

    // Add user on context if we have an authorization header
    try {
      const authorization = req.headers.authorization;
      const token = authorization ? authorization.split(' ')[1] : '';
      if (token) {
        // if we have a token, verify it
        if (config.FEDERATED) {
          // when federated, let the gateway verify the token
          context.user = decode(token) as JWTPayload;
        } else {
          context.user = verify(token, config.ACCESS_TOKEN_SECRET) as JWTPayload;
        }
      }
    } catch (err) {
      // we don't throw an error here
      // the auth directives will handle the authorization
      log.warn('Failed to verify accessToken, ERR: ', err.message);
    }
    return context;
  },
  dataSources() {
    return {
      user: new UserDataSource(),
    };
  },
  formatError: (err) => {
    log.error('GRAPQH_ERROR >>>', util.inspect(err, false, 4, true /* enable colors */));
    return err;
  },
  playground: true,
  introspection: true,
});

export const schema = graphqlSchema;
