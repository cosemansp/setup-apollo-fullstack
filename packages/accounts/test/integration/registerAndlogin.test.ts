import gql from 'graphql-tag';
import { createTestClient } from '@test/apolloServerTesting';
import { apolloServer } from '@/graphql/apolloServer';
import { openDb, closeDb, clearDb } from '@test/dbHelper';
import { UserModel } from '@/domain/userModel';

const { mutate } = createTestClient(apolloServer);

const REGISTER_MUTATION = gql`
  mutation registerUser($input: RegisterInput!) {
    register(input: $input) {
      __typename
      ... on Registered {
        user {
          id
          name
          email
        }
        accessToken
      }
      ... on UserAlreadyExistError {
        reason
      }
    }
  }
`;

const LOGIN_MUTATION = gql`
  mutation loginUser($input: LoginInput!) {
    login(input: $input) {
      __typename
      ... on Authentication {
        user {
          id
          name
          email
        }
        accessToken
      }
      ... on NotAuthorizedError {
        reason
      }
    }
  }
`;

describe('registerAndLogin', () => {
  beforeAll(() => openDb());
  beforeEach(clearDb);
  afterAll(closeDb);

  test('scenario register & login', async () => {
    // register
    const registerResult = await mutate({
      mutation: REGISTER_MUTATION,
      variables: {
        input: {
          name: 'John Doe',
          email: 'john.doe@gmail.com',
          password: '12345',
        },
      },
    });
    expect(registerResult.errors).toBeUndefined();
    expect(registerResult.data).toMatchSnapshot({
      register: {
        user: {
          id: expect.any(String),
        },
        accessToken: expect.any(String),
      },
    });

    // success login
    const loginResult = await mutate({
      mutation: LOGIN_MUTATION,
      variables: {
        input: {
          email: 'john.doe@gmail.com',
          password: '12345',
        },
      },
    });
    expect(loginResult.errors).toBeUndefined();
    expect(loginResult.data).toMatchSnapshot({
      login: {
        user: {
          id: expect.any(String),
        },
        accessToken: expect.any(String),
      },
    });

    // bad login
    const loginResult2 = await mutate({
      mutation: LOGIN_MUTATION,
      variables: {
        input: {
          email: 'john.doe@gmail.com',
          password: 'unknown',
        },
      },
    });
    expect(loginResult2.errors).toBeUndefined();
    expect(loginResult2.data).toEqual({
      login: {
        __typename: 'NotAuthorizedError',
        reason: expect.any(String),
      },
    });
  });

  test(`register fails when email exist`, async () => {
    const user = new UserModel({
      name: 'John Doe',
      email: 'john.doe@gmail.com',
    });
    await user.save();

    // register
    const registerResult = await mutate({
      mutation: REGISTER_MUTATION,
      variables: {
        input: {
          name: 'John Doe',
          email: 'john.doe@gmail.com',
          password: '12345',
        },
      },
    });
    expect(registerResult.errors).toBeUndefined();
    expect(registerResult.data).toEqual({
      register: {
        __typename: 'UserAlreadyExistError',
        reason: expect.any(String),
      },
    });
  });
});
