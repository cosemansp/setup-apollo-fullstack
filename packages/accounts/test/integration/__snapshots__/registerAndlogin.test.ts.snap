// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerAndLogin scenario register & login 1`] = `
Object {
  "register": Object {
    "__typename": "Registered",
    "accessToken": Any<String>,
    "user": Object {
      "email": "john.doe@gmail.com",
      "id": Any<String>,
      "name": "John Doe",
    },
  },
}
`;

exports[`registerAndLogin scenario register & login 2`] = `
Object {
  "login": Object {
    "__typename": "Authentication",
    "accessToken": Any<String>,
    "user": Object {
      "email": "john.doe@gmail.com",
      "id": Any<String>,
      "name": "John Doe",
    },
  },
}
`;
